[Dialect/Plan] Update Plan dialect to use non-dps calling convention

jhalakpatel opened the pull request:
The Plan dialect has been enhanced to support non-DPS calling conventions, with updates to PlanAllocTensorsPass and CreateClosedRegionsPass. Additionally, the Plan transformation now converts both non-DPS and DPS group operations to CallAllocOp and CallOp respectively, required for subsequent transformations.

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Compiler/StableHloToExecutable.h:
```diff
@@ -128,6 +128,9 @@ struct StableHLOToExecutableOptions : public mlir::OptionsContext {
   /// Whether to disallow host tensors in TensorRT clusters.
   bool disallowHostTensorsInTensorRTClusters = false;
 
+  /// Whether to use non-DPS style calling convention.
```
This should say something along the lines of "use non-DPS calling convention for the entrypoint function and backend types that support allocating results"

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/IR/PlanOps.td:
```diff
@@ -255,16 +273,75 @@ def Plan_InlineClosedGroupOp : Plan_GroupOpBase<"inline_closed_group", [
+// InlineClosedGroupNonDPSOp
+//===----------------------------------------------------------------------===//
+
+def Plan_InlineClosedGroupNonDPSOp : Plan_InlineClosedGroupBase<"inline_closed_group_non_dps", [
```
Let's use a name similar to the pattern used in the TensorRT ops which you are adding, e.g. 

```suggestion
def Plan_InlineClosedAllocGroupOp : Plan_InlineClosedGroupBase<"inline_closed_alloc_group", [
```

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/AllocTensors.cpp:
```diff
@@ -856,11 +856,13 @@ class AllocTensorsPass
-    if (failed(rewriteNotPrivateFuncsToDPS(rewriter, op))) {
-      op->emitError("Failed to convert non-private functions to DPS");
-      return signalPassFailure();
+    if (!useNonDPSCallConv) {
```
Missing test for `plan-alloc-tensors`,

You'll need to add a `RUN` line in `test/Dialect/Plan/plan-alloc-tensors.mlir` where this option is invoked with `true` and add `CHECK` statements to show what is the desired result.

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/OutlineClusters.cpp:
```diff
@@ -292,8 +302,13 @@ createFunctionsFromRegions(RewriterBase &rewriter, Region &region,
+    if (auto group = dyn_cast<plan::InlineClosedGroupOp>(op)) {
+      if (failed(outlineTensorRTRegion(rewriter, group)))
+        return WalkResult::interrupt();
+      return WalkResult::advance();
```
Don't use `else` after `return` 

https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/OutlineClusters.cpp:
```diff
@@ -264,6 +266,13 @@ static LogicalResult outlineTensorRTRegion(RewriterBase &rewriter,
 
+static LogicalResult outlineTensorRTRegion(RewriterBase &rewriter,
+                                           plan::InlineClosedGroupNonDPSOp op) {
+  llvm_unreachable("outlinining non-dps inline closed group ops to tensorrt "
```
when `llvm_unreachable` is used, it results in a crash/abort, which is not desired. Error should be propagated to callers so that the pass can signal failure. Just `return op->emitError(.....)` should be sufficient here.

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/Transforms/Passes.td:
```diff
@@ -248,6 +248,9 @@ def StablehloClusteringPass : Pass<"stablehlo-clustering", "::mlir::ModuleOp"> {
       " on all functions">,
+    Option<"useNonDPSCallConv",
+      "use-non-dps-call-conv", "bool", "false",
+      "allow tensorrt based output allocations using output allocator">,      
```
Should be "allow backend clusters to directly allocate outputs"

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/Transforms/Passes.td:
```diff
@@ -332,7 +335,10 @@ def CreateClosedRegionsPass : Pass<"plan-create-closed-regions", "::mlir::Module
+      "to traversal order">,
+    Option<"useNonDPSCallConv", "use-non-dps-call-conv", "bool", 
+           /*default=*/"false", 
+           "Allow TensorRT-based output allocations using output allocator">
```
Should be "allow backend clusters to directly allocate outputs"

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/Transforms/Passes.td:
```diff
@@ -428,6 +434,13 @@ def PlanAllocTensorsPass : Pass<"plan-alloc-tensors",
+  let options = [
+    Option<"useNonDPSCallConv", "use-non-dps-call-conv", "bool", 
+           /*default=*/"false", 
+           "Allow TensorRT-based output allocations using output allocator">
```
Should be "allow backend clusters to directly allocate outputs"

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -351,70 +351,11 @@ static void remapAnalysisState(DataFlowSolver &solver, ValueRange originals,
   }
 }
```
You have lots of changes to this pass, so there the test `test/Dialect/Plan/create-closed-regions.mlir` should also be updated/expanded.

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/Passes.cpp:
```diff
@@ -73,6 +76,11 @@ struct ClusteringPipelineCliOpts
+  Option<bool> useNonDPSCallConv{
+      *this, "use-non-dps-call-conv",
+      llvm::cl::desc(
+          "allow tensorrt based output allocations using output allocator"),
```
Please correct doc comment throughout similar to other comments for this option.

christopherbate reviewed (changes requested):
Overall, it looks good, but it is missing updates to tests for the two main passes that were changed (`alloc-tensors` and `create-closed-regions`).

In addition, there should be new tests in 

- `test/Dialect/Plan/invalid.mlir` 
-  `test/Dialect/Plan/roundtrip.mlir` 

Specifically for the new operation that was added.

jhalakpatel commented:
> Overall, it looks good, but it is missing updates to tests for the two main passes that were changed (`alloc-tensors` and `create-closed-regions`).
> 
> In addition, there should be new tests in
> 
> * `test/Dialect/Plan/invalid.mlir`
> * `test/Dialect/Plan/roundtrip.mlir`
> 
> Specifically for the new operation that was added.

@christopherbate I have addressed your comments. Please review again.

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/IR/PlanOps.td:
```diff
@@ -255,16 +273,75 @@ def Plan_InlineClosedGroupOp : Plan_GroupOpBase<"inline_closed_group", [
+  The `plan.inline_closed_alloc_group` operation is a variant of the
+  `plan.inline_closed_group` operation that does not use destination-passing style
+  (DPS). It is isolated from above and explicitly captures input operands,
+  but unlike its DPS counterpart, it does not capture destination operands.
```
```suggestion
  but unlike its DPS counterpart, it does not capture destination operands because its results must be lowered to allocation(s). The allocations may  or may not be of a size that can only be computed inside of the region.
```

christopherbate commented on mlir-tensorrt/compiler/include/mlir-tensorrt/Dialect/Plan/IR/PlanOps.td:
```diff
@@ -255,16 +273,75 @@ def Plan_InlineClosedGroupOp : Plan_GroupOpBase<"inline_closed_group", [
 
+  %2 = plan.inline_closed_alloc_group target(#plan.cluster_target<tensorrt>)
+    inputs(%0, %1 : tensor<?xf32>, index)
+    in_attrs [#plan.bounds<shape, , >, #plan.bounds<none>] -> tensor<?xf32> {
```
This syntax looks incorrect: `#plan.bounds<shape, , >`, did you mean to fill in something there?

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/IR/PlanOps.cpp:
```diff
@@ -465,6 +467,68 @@ void InlineClosedGroupOp::build(OpBuilder &b, OperationState &state,
+LogicalResult InlineClosedAllocGroupOp::verify() {
+  Operation *op = getOperation();
+  // Check for res_attrs
+  if (op->hasAttr("res_attrs")) {
```
Nit: remove braces on single-statement bodies of `if` statements 

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/IR/PlanOps.cpp:
```diff
@@ -465,6 +467,68 @@ void InlineClosedGroupOp::build(OpBuilder &b, OperationState &state,
+    Region &region, OpAsmSetValueNameFn setNameFn) {
+  assert(region.front().getNumArguments() == getInputs().size() &&
+         "expected one block arg for each input argument");
+  unsigned numInputs = getInputs().size();
```
Not really a need for the assertion here. The code here isn't going to crash if the assertion fails. `numInputs` will be an unused variable when assertions are turned off.

I believe that `region.front().getArguments()` can also be replaced with just `region.getArguments()`. A Region's arguments are the arguments of the entry block.

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -351,70 +351,11 @@ static void remapAnalysisState(DataFlowSolver &solver, ValueRange originals,
-  rewriter.replaceOp(op, replacements);
+static FailureOr<SmallVector<BoundsAttr>>
+getInputAttributes(RewriterBase &rewriter, DataFlowSolver &solver, Location loc,
+                   SmallVector<Value> const &inputs) {
```
```suggestion
                  ValueRange inputs) {
```

In general, use `ValueRange` and `TypeRange` to represent const range over values and types, respectively. For other cases, use `ArrayRef<T>` rather than `const SmallVector<T> &`. For mutable SmallVector, use `SmallVectorImpl<T> &` to allow for different implicit size parameters. See note here https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h.

jhalakpatel replied:
Thanks a lot! This is super helpful.

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -425,30 +366,16 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
-      return closedGroupOp->emitOpError("input operand #")
-             << idx << " of type " << input.getType()
+      return emitError(loc)
+             << ("input operand #") << idx << " of type " << input.getType()
```
Nit: parentheses can be removed here

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+    FailureOr<DestinationOperandMaterializationResult> destResult =
+        materializeDestinationOperand(rewriter, res.getLoc(), op,
+                                      res.getResultNumber(), solver);
+    if (failed(destResult)) {
```
Nit: remove curly braces for single-statement bodies of `if` statements 
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+// Helper function to create DPS closed group op.
+static LogicalResult createInlineClosedGroupOp(
+    RewriterBase &rewriter, plan::InlineGroupOp op, DataFlowSolver &solver,
+    const SmallVector<Value> &inputs,
```
```suggestion
   ValueRange inputs,
```

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+static LogicalResult createInlineClosedGroupOp(
+    RewriterBase &rewriter, plan::InlineGroupOp op, DataFlowSolver &solver,
+    const SmallVector<Value> &inputs,
+    const SmallVector<DestinationOperandMaterializationResult>
```
Please use

```
    ArrayRef<DestinationOperandMaterializationResult>
        destinationOperands) {
```

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+  SmallVector<Value> replacements;
+  replacements.reserve(destinationOperands.size());
+  for (auto [newResult, destOperand, originalType] :
+       llvm::zip(closedGroupOp->getResults(), destinationOperands,
```
Use `zip_equal` if all lists are supposed to be the same size (and it will assert as such). Otherwise the range is limited by the smallest element.

christopherbate commented on mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp:
```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+  closedGroupOp.setResAttrsAttr(resultAttrs);
+
+  // Create the closed region input profilw attrs.
+  auto inputAttr = getInputAttributes(rewriter, solver, closedGroupOp->getLoc(),
```
Replace `auto` with actual typename. `auto` is only encouraged when type is obvious (e.g. due to template parameter on RHS of `=`) or very long/complicated (e.g. specialized iterator types).

See https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

christopherbate commented on mlir-tensorrt/test/Dialect/Plan/create-closed-regions.mlir:
```diff
@@ -59,6 +69,21 @@ func.func @test_simple_shape_bound(%arg0: tensor<?x10xf32> {tensorrt.shape_profi
+//       CHECK-ALLOC:       yield %[[v4]] : tensor<?x10xf32>
+//       CHECK-ALLOC:     return %[[v2]] : tensor<?x10xf32>
+
 // -----
```
Include at least a `CHECK-ALLOC-LABEL` for the next test. Otherwise you don't know whether the `CHECK-ALLOC` statements in this last test are actually bleeding through the end of the test and matching the result of some other test segment. It is a common source of false positives that go undetected in large test files with multiple FileCheck prefixes in use.

christopherbate commented on mlir-tensorrt/test/Dialect/Plan/plan-alloc-tensors.mlir:
```diff
@@ -232,6 +303,12 @@ func.func @test_return_arg(%arg0: tensor<10xf32>) -> tensor<10xf32> {
 
+// CHECK-ALLOC-LABEL: @test_return_arg
+//  CHECK-ALLOC-SAME: (%[[arg0:.+]]: tensor<10xf32>)
+//       CHECK-ALLOC:     return %[[arg0]]
```
```suggestion
//  CHECK-ALLOC-NEXT:     return %[[arg0]]

```

christopherbate reviewed (approved):
Some additional more minor comments, otherwise LGTM.
